File: min-camlMy/syntax.ml
open Lexing

type t = (* MinCamlの構文を表現するデータ型 (caml2html: syntax_t) *)
  | Unit
  | Bool of bool
  | Int of int
  | Float of float
  | Not of t
  | Neg of t
  | Add of t * t
  | Sub of t * t
  | SLL of t * t	(* 第二オペランドは即値のみ *)
  | SRA of t * t	(* 同上 *)
  | FNeg of t
  | FAdd of t * t
  | FSub of t * t
  | FMul of t * t
  | FDiv of t * t
  | Eq of t * t
  | LE of t * t
  | If of t * t * t
  | Let of (Id.t * Type.t) * t * t	(* let *)
  | Var of Id.t
  | LetRec of fundef * t
  | App of t * t list		(* この時点では t == Var Id.t かな*)
  | Tuple of t list
  | LetTuple of (Id.t * Type.t) list * t * t
  | Array of t * t
  | Get of t * t
  | Put of t * t * t
and fundef = { name : Id.t * Type.t; args : (Id.t * Type.t) list; body : t }


type t' = (* positionを記憶 *)
  | Unit'
  | Bool' of bool
  | Int' of int
  | Float' of float
  | Not' of t' * position
  | Neg' of t' * position
  | Add' of t' * t' * position
  | Sub' of t' * t' * position
  | SLL' of t' * t' * position
  | SRA' of t' * t' * position
  | FNeg' of t' * position
  | FAdd' of t' * t' * position
  | FSub' of t' * t' * position
  | FMul' of t' * t' * position
  | FDiv' of t' * t' * position
  | Eq' of t' * t' * position
  | LE' of t' * t' * position
  | If' of t' * t' * t' * position
  | Let' of (Id.t * Type.t) * t' * t' * position
  | Var' of Id.t
  | LetRec' of fundef' * t' * position
  | App' of t' * t' list * position
  | Tuple' of t' list * position
  | LetTuple' of (Id.t * Type.t) list * t' * t' * position
  | Array' of t' * t' * position
  | Get' of t' * t' * position
  | Put' of t' * t' * t' * position
and fundef' = { name' : Id.t * Type.t; args' : (Id.t * Type.t) list; body' : t' }

File: hs-mincaml/Syntax.hs
module Syntax where

import qualified Type as T
import qualified IdMod as I

data T = Unit
       | Bool Bool
       | Int Int
       | Float Float
       | Not T
       | Neg T
       | AddP T T               -- polymorphism
       | Add T T | Sub T T
       | SLL T Int | SRA T Int
       | FNeg T
       | Fabs T
       | FAdd T T | FSub T T | FMul T T | FDiv T T
       | Eq T T | Le T T
       | Sqrt T
       | If T T T
       | Var I.Id
       | Let (I.Id,T.T)  T T                          
       | LetRec Fundef T
       | App T [T]
       | Tuple [T]
       | LetTuple [(I.Id, T.T)] T T
       | Array T T
       | Get T T
       | Put T T T      
       deriving (Eq, Ord)
                
data Fundef = Fundef {name :: (I.Id, T.T), args :: [(I.Id, T.T)], body :: T} 
            deriving (Eq, Ord)
                     
instance Show T where
  show exp = printSyntax exp 0
    
printSyntax :: T -> Int -> String
printSyntax exp dep = 
  indent ++ 
  (case exp of
    Unit        -> "Unit"++"\n"
    Bool b      -> "Bool: "++(show b)++"\n"
    Int i       -> "Int: "++(show i)++"\n"
    Float f     -> "Float: "++(show f)++"\n"
    Not e       -> "Not\n"++(printE e)
    Neg e       -> "Neg\n"++(printE e)
    FNeg e      -> "FNeg\n"++(printE e)
    Fabs e      -> "Fabs\n"++(printE e)
    Sqrt e      -> "Sqrt\n"++(printE e)    
    AddP e1 e2   -> "AddP\n"++(printE e1)++(printE e2)    
    Add e1 e2   -> "Add\n"++(printE e1)++(printE e2)
    Sub e1 e2   -> "Sub\n"++(printE e1)++(printE e2)
    SLL e i     -> "SLL: "++(show i)++"\n"++(printE e)
    SRA e i     -> "SRA: "++(show i)++"\n"++(printE e)
    FAdd e1 e2  -> "FAdd\n"++(printE e1)++(printE e2)
    FSub e1 e2  -> "FSub\n"++(printE e1)++(printE e2)
    FMul e1 e2  -> "FMul\n"++(printE e1)++(printE e2)
    FDiv e1 e2  -> "FDiv\n"++(printE e1)++(printE e2)
    Eq e1 e2    -> "Eq\n"++(printE e1)++(printE e2)
    Le e1 e2    -> "Le\n"++(printE e1)++(printE e2)
    If e1 e2 e3 -> "If\n"++(printE e1)++
                   indent++"Then\n"++(printE e2)++
                   indent++"Else\n"++(printE e3)
    Var x       -> "Var: "++x++"\n"
    Let xt e1 e2     -> "Let: "++(printXT xt)++"\n"++(printE e1)++
                        indent++"In\n"++(printE e2)
    LetRec (Fundef{name=xt, args=yts, body=e1}) e2
      -> "LetRec: "++(printXT xt)++" args:"++(concatMap printXT yts)++"\n"
         ++(printE e1)++indent++"In\n"++(printE e2)
    App e es    -> "App\n"++(printE e)++(concatMap printE es)
    Tuple es    -> "Tuple\n"++(concatMap printE es)
    LetTuple xts e1 e2  -> "LetTuple: "++(concatMap printXT xts)++"\n"
                           ++(printE e1)++indent++"In\n"++(printE e2)
    Array e1 e2 -> "Array\n"++(printE e1)++(printE e2)
    Get e1 e2   -> "Get\n"++(printE e1)++(printE e2)
    Put e1 e2 e3-> "Put\n"++(printE e1)++(printE e2)++(printE e3)
  )
    where indent = (concat $ take dep (repeat " "))
          printE = \e -> printSyntax e (dep+1)
          printXT (x,t) = "["++x++" ("++(show t)++")] "
